scripts/workarounds/boot-order/metal-lib.sh

```diff
@@ -0,0 +1,510 @@
+    index=0
+    [ -z "$name" ] && name='CRAY Linux' # Note: if CRAY Linux is observed then the customer has installed a non-SLES distro.
+    for disk in $(mdadm --detail $(blkid -L ${boot_authority}) | grep /dev/sd | awk '{print $NF}'); do
+        # Add '--suse-enable-tpm' to grub2-install once we need TPM.
+        grub2-install --no-rs-codes --suse-force-signed --root-directory $working_path --removable "$disk"
+        efibootmgr -c -D -d "$disk" -p 1 -L "CRAY UEFI OS $index" -l '\efi\boot\bootx64.efi' | grep -i cray
+        index=$((index + 1))
+    done
+
+    # Get the kernel command we used to boot.
+    init_cmdline=$(cat /proc/cmdline)
+    disk_cmdline=''
+    for cmd in $init_cmdline; do
+        # cleans up first argument when running this script on a disk-booted system
+        if [[ $cmd =~ kernel$ ]]; then
+            cmd=$(basename "$(echo $cmd  | awk '{print $1}')")
+        fi
+        if [[ $cmd =~ ^rd.live.overlay.reset ]] ; then :
+        elif [[ $cmd =~ ^rd.debug ]] ; then :
+        # removes all metal vars, and escapes anything that iPXE was escaping
+        # metal vars are used for customizing nodes on deployment, they don't need
+        # to stick around for runtime.
+        # (i.e. ds=nocloud-net;s=http://$url will get the ; escaped)
+        # removes netboot vars
+        elif [[ ! $cmd =~ ^metal. ]] && [[ ! $cmd =~ ^ip=.*:dhcp ]] && [[ ! $cmd =~ ^bootdev= ]]; then
+            disk_cmdline="$(trim $disk_cmdline) ${cmd//;/\\;}"
+        fi
+    done
+
+    # ensure no-wipe is now set for disk-boots.
+    disk_cmdline="$disk_cmdline metal.no-wipe=1"
+
+    # Make our grub.cfg file.
+    # TODO: Add disk-rebuild option
+    cat << EOF > $working_path/boot/grub2/grub.cfg
+set timeout=10
+set default=0 # Set the default menu entry
+menuentry "$name" --class sles --class gnu-linux --class gnu {
+    set gfxpayload=keep
+    # needed for compression
+    insmod gzio
+    # needed for partition manipulation
+    insmod part_gpt
+    # needed for block device handles
+    insmod diskfilter
+    # needed for RAID (this does not always load despite this entry)
+    insmod mdraid1x
+    # verbosely define accepted formats (ext2/3/4 & xfs)
+    insmod ext2
+    insmod xfs
+    echo    'Loading kernel ...'
+    linuxefi \$prefix/../$disk_cmdline
+    echo    'Loading initial ramdisk ...'
+    initrdefi \$prefix/../$INITRD
+}
+EOF
+}
+
+function update_auxiliary_fstab {
+    local working_path=/metal/recovery
+    mkdir -pv $working_path
+
+    # Mount at boot
+    if [ -f /etc/fstab.metal ] && grep -q "${boot_authority}" /etc/fstab.metal; then :
+    else
+        printf "# added by cloud-init: \n% -18s\t% -18s\t%s\t%s %d %d\n" "${boot_scheme}=${boot_authority}" $working_path vfat defaults 0 0 >> /etc/fstab.metal
+    fi
+}
+
+function get_boot_artifacts {
+    local squashfs_storage
+    # TODO: Test using the initrd from /boot, knowing it won't match what was in S3/apache2.
```

Comment:
Todos would be better captured in Jiras than in comments in scripts that are visible to customers, IMO
```suggestion
```

Reply:
It pains me but I'm going to omit amending the metal-lib.sh script with changes to keep the diff to the original script in 1.3/1.4 the same.